### PR TITLE
Add flow monitoring feature implementation

### PR DESCRIPTION
**Pull request by rozdy**
Related to #3968

### CODE REVIEW

File: src-java/flowmonitoring-topology/flowmonitoring-storm-topology/src/main/java/org/openkilda/wfm/topology/flowmonitoring/bolts/TickBolt.java [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+        }
+        if (currentTimeMillis > this.lastCheckTimestamp + flowSlaCheckInterval) {
+            this.lastCheckTimestamp = currentTimeMillis;
+            emit(input, new Values());
+        }
+    }
+
+    @Override
+    public void declareOutputFields(OutputFieldsDeclarer declarer) {
+        declarer.declare(new Fields());
```
Comment (timofei-durakov) [inline src-java/flowmonitoring-topology/flowmonitoring-storm-topology/src/main/java/org/openkilda/wfm/topology/flowmonitoring/bolts/TickBolt.java:50]: that's wierd

File: src-java/flowmonitoring-topology/flowmonitoring-storm-topology/src/main/java/org/openkilda/wfm/topology/flowmonitoring/FlowMonitoringTopology.java [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+    }
+
+    private void coordinatorSpout(TopologyBuilder topologyBuilder) {
+        declareSpout(topologyBuilder, new CoordinatorSpout(), ComponentId.COORDINATOR_SPOUT.name());
+    }
+
+    private void tickBolt(TopologyBuilder topologyBuilder) {
+        declareBolt(topologyBuilder, new TickBolt(topologyConfig.getFlowSlaCheckIntervalSeconds()),
+                ComponentId.TICK_BOLT.name())
+                .allGrouping(ComponentId.COORDINATOR_SPOUT.name());
```
Comment (timofei-durakov) [inline src-java/flowmonitoring-topology/flowmonitoring-storm-topology/src/main/java/org/openkilda/wfm/topology/flowmonitoring/FlowMonitoringTopology.java:95]: why do you need coordinator?

File: src-java/flowmonitoring-topology/flowmonitoring-storm-topology/src/main/java/org/openkilda/wfm/topology/flowmonitoring/mapper/FlowMapper.java [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+    @Mapping(target = "maxLatency", source = "maxLatency")
+    FlowState toFlowState(Flow flow);
+
+    @Mapping(target = "forwardPath", source = "info.flowPath.forwardPath")
+    @Mapping(target = "reversePath", source = "info.flowPath.reversePath")
+    @Mapping(target = "maxLatency", source = "maxLatency")
+    FlowState toFlowState(UpdateFlowInfo info);
+
+    default List<Link> toLinks(List<PathSegment> pathSegments) {
+        return pathSegments.stream().map(this::toLink).collect(Collectors.toList());
```
Comment (timofei-durakov) [inline src-java/flowmonitoring-topology/flowmonitoring-storm-topology/src/main/java/org/openkilda/wfm/topology/flowmonitoring/mapper/FlowMapper.java:51]: add check for null just in case here

File: src-java/flowmonitoring-topology/flowmonitoring-storm-topology/src/main/java/org/openkilda/wfm/topology/flowmonitoring/model/LinkState.java [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+@Data
+@Builder
+public class LinkState {
+
+    private long rttLatency;
+    private long rttTimestamp;
+    private long oneWayLatency;
+
+    /**
+     * Get most accurate latency for the link.
```
Comment (timofei-durakov) [inline src-java/flowmonitoring-topology/flowmonitoring-storm-topology/src/main/java/org/openkilda/wfm/topology/flowmonitoring/model/LinkState.java:30]: I'd reword this, as I understand it's not about accuracy, but about the freshness of the metric

File: src-java/flowmonitoring-topology/flowmonitoring-storm-topology/src/main/java/org/openkilda/wfm/topology/flowmonitoring/service/ActionService.java [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+public class ActionService {
+
+    /**
+     * Check flow SLA is violated.
+     */
+    public void checkFlowSla(String flowId, FlowDirection direction, long latency, long maxLatency) {
+        if (latency > maxLatency) {
+            long latencyMs = latency / 1000;
+            long maxLatencyMs = maxLatency / 1000;
+            log.warn("{} path for flow '{}' violates latency SLA {}/{} ms",
```
Comment (timofei-durakov) [inline src-java/flowmonitoring-topology/flowmonitoring-storm-topology/src/main/java/org/openkilda/wfm/topology/flowmonitoring/service/ActionService.java:32]: maybe it would be better to have adaptive logging here, I would prefer to have a warn once in a minute at least, and all other triggers to be info for example

File: confd/vars/main.yaml [hunk 6]
```diff
@@ -161,6 +161,8 @@ kilda_storm_parallelism_level: 1
 kilda_storm_flow_hs_parallelism: 10
 kilda_storm_parallelism_workers_count: 1
 
+kilda_storm_flow_monitoring_parallelism: 1
```
Comment (timofei-durakov) [inline confd/vars/main.yaml:164]: is it enough?

File: src-java/flowmonitoring-topology/flowmonitoring-messaging/src/main/java/org/openkilda/messaging/info/flow/UpdateFlowInfo.java [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+    private static final long serialVersionUID = 1L;
+
+    @JsonProperty("flow_id")
+    protected String flowId;
+
+    @JsonProperty("flow_path")
+    protected FlowPathDto flowPath;
+
+    @JsonProperty("max_latency")
+    protected Long maxLatency;
```
Comment (timofei-durakov) [inline src-java/flowmonitoring-topology/flowmonitoring-messaging/src/main/java/org/openkilda/messaging/info/flow/UpdateFlowInfo.java:43]: why is it Object not a primitive?

**Review by timofei-durakov — CHANGES_REQUESTED** [review]

**Review by niksv — CHANGES_REQUESTED** [review]

File: src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/fsm/create/FlowCreateFsm.java [hunk 8]
```diff
@@ -457,10 +460,19 @@ protected String getCrudActionName() {
...
+                    .to(State.NOTIFY_FLOW_MONITOR)
+                    .on(Event.NEXT)
+                    .perform(new OnFinishedAction(dashboardLogger));
+            builder.transition()
+                    .from(State.FINISHED_WITH_ERROR)
+                    .to(State.NOTIFY_FLOW_MONITOR)
+                    .on(Event.NEXT)
+                    .perform(new OnFinishedWithErrorAction(dashboardLogger));
+
+            builder.defineFinalState(State.NOTIFY_FLOW_MONITOR)
```
[the remark this answers is not in the document]
Comment (dpoltavets) [reply src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/fsm/create/FlowCreateFsm.java:474]: +1

File: src-java/flowmonitoring-topology/flowmonitoring-messaging/src/main/java/org/openkilda/messaging/info/flow/UpdateFlowInfo.java (hunk 7 above)
Comment (dpoltavets) [reply src-java/flowmonitoring-topology/flowmonitoring-messaging/src/main/java/org/openkilda/messaging/info/flow/UpdateFlowInfo.java:43]: What about the `max_latency_tier2` field?

File: src-java/flowmonitoring-topology/flowmonitoring-storm-topology/src/main/java/org/openkilda/wfm/topology/flowmonitoring/bolts/IslCacheBolt.java [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+        this.islRttLatencyExpiration = islRttLatencyExpiration;
+    }
+
+    @PersistenceContextRequired(requiresNew = true)
+    protected void init() {
+        islCacheService = new IslCacheService(persistenceManager, islRttLatencyExpiration);
+    }
+
+    @Override
+    protected void handleInput(Tuple input) throws PipelineException {
```
[the remark this answers is not in the document]
Comment (dpoltavets) [reply src-java/flowmonitoring-topology/flowmonitoring-storm-topology/src/main/java/org/openkilda/wfm/topology/flowmonitoring/bolts/IslCacheBolt.java:70]: +1

**Review by dpoltavets — CHANGES_REQUESTED** [review]

File: src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/fsm/create/FlowCreateFsm.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (rozdy) [reply src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/fsm/create/FlowCreateFsm.java:474]: To swap them we need to introduce two notify states (for success and fail scenarios) or some other way to handle different finish action. For now I don't have better solution. What do you think?

File: src-java/flowmonitoring-topology/flowmonitoring-storm-topology/src/main/java/org/openkilda/wfm/topology/flowmonitoring/FlowMonitoringTopologyConfig.java [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+        return getKafkaTopics().getNetworkFlowMonitoringNotifyTopic();
+    }
+
+    @Key("flow.sla.check.interval.seconds")
+    @Default("30")
+    int getFlowSlaCheckIntervalSeconds();
+
+    @Key("flow.rtt.stats.expiration")
+    @Default("3")
+    int getFlowRttStatsExpiration();
```
[the remark this answers is not in the document]
Comment (rozdy) [reply src-java/flowmonitoring-topology/flowmonitoring-storm-topology/src/main/java/org/openkilda/wfm/topology/flowmonitoring/FlowMonitoringTopologyConfig.java]: Renamed.

File: src-java/flowmonitoring-topology/flowmonitoring-storm-topology/src/main/java/org/openkilda/wfm/topology/flowmonitoring/bolts/TickBolt.java (hunk 1 above)
Comment (rozdy) [reply src-java/flowmonitoring-topology/flowmonitoring-storm-topology/src/main/java/org/openkilda/wfm/topology/flowmonitoring/bolts/TickBolt.java]: Reworked.

File: src-java/flowmonitoring-topology/flowmonitoring-storm-topology/src/main/java/org/openkilda/wfm/topology/flowmonitoring/FlowMonitoringTopology.java (hunk 2 above)
Comment (rozdy) [reply src-java/flowmonitoring-topology/flowmonitoring-storm-topology/src/main/java/org/openkilda/wfm/topology/flowmonitoring/FlowMonitoringTopology.java]: Reworked.

File: src-java/flowmonitoring-topology/flowmonitoring-storm-topology/src/main/java/org/openkilda/wfm/topology/flowmonitoring/mapper/FlowMapper.java (hunk 3 above)
Comment (rozdy) [reply src-java/flowmonitoring-topology/flowmonitoring-storm-topology/src/main/java/org/openkilda/wfm/topology/flowmonitoring/mapper/FlowMapper.java:60]: Added.

File: src-java/flowmonitoring-topology/flowmonitoring-storm-topology/src/main/java/org/openkilda/wfm/topology/flowmonitoring/model/LinkState.java (hunk 4 above)
Comment (rozdy) [reply src-java/flowmonitoring-topology/flowmonitoring-storm-topology/src/main/java/org/openkilda/wfm/topology/flowmonitoring/model/LinkState.java]: Fixed.

File: src-java/flowmonitoring-topology/flowmonitoring-storm-topology/src/main/java/org/openkilda/wfm/topology/flowmonitoring/service/ActionService.java (hunk 5 above)
Comment (rozdy) [reply src-java/flowmonitoring-topology/flowmonitoring-storm-topology/src/main/java/org/openkilda/wfm/topology/flowmonitoring/service/ActionService.java:33]: It's for testing purpose. This section will be covered in next patches on flow SLA reactions.

File: confd/vars/main.yaml (hunk 6 above)
Comment (rozdy) [reply confd/vars/main.yaml:164]: Current flow and isl caches are implemented without parallelism support.

File: src-java/flowmonitoring-topology/flowmonitoring-messaging/src/main/java/org/openkilda/messaging/info/flow/UpdateFlowInfo.java (hunk 7 above)
Comment (rozdy) [reply src-java/flowmonitoring-topology/flowmonitoring-messaging/src/main/java/org/openkilda/messaging/info/flow/UpdateFlowInfo.java:45]: Because we can have a flows with null maxLatency.
Added maxLatencyTier2.

File: src-java/flowmonitoring-topology/flowmonitoring-messaging/src/main/java/org/openkilda/messaging/info/flow/UpdateFlowInfo.java [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonInclude;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.fasterxml.jackson.databind.annotation.JsonSerialize;
+import lombok.Value;
+
+/**
+ * Represents update flow info.
+ */
+@Value
```
[the remark this answers is not in the document]
Comment (rozdy) [reply src-java/flowmonitoring-topology/flowmonitoring-messaging/src/main/java/org/openkilda/messaging/info/flow/UpdateFlowInfo.java:31]: Done.

File: docker/zookeeper/bin/create-zknode-and-start.sh [hunk 12]
```diff
@@ -77,6 +77,13 @@ done
 /opt/zookeeper/bin/zkCli.sh -server 127.0.0.1:2181 create /${KILDA_ZKNODE}/flowhs/blue/state ""
 /opt/zookeeper/bin/zkCli.sh -server 127.0.0.1:2181 create /${KILDA_ZKNODE}/flowhs/blue/build-version "v3r\$i0n"
 
+# add flowmonitoring topology zkNodes
+/opt/zookeeper/bin/zkCli.sh -server 127.0.0.1:2181 create /${KILDA_ZKNODE}/flowmonitoring ""
+/opt/zookeeper/bin/zkCli.sh -server 127.0.0.1:2181 create /${KILDA_ZKNODE}/flowmonitoring/blue ""
+/opt/zookeeper/bin/zkCli.sh -server 127.0.0.1:2181 create /${KILDA_ZKNODE}/flowmonitoring/blue/signal "START"
+/opt/zookeeper/bin/zkCli.sh -server 127.0.0.1:2181 create /${KILDA_ZKNODE}/flowmonitoring/blue/state ""
+/opt/zookeeper/bin/zkCli.sh -server 127.0.0.1:2181 create /${KILDA_ZKNODE}/flowmonitoring/blue/build-version "v3r\$i0n"
```
[the remark this answers is not in the document]
Comment (rozdy) [reply docker/zookeeper/bin/create-zknode-and-start.sh:85]: Done.

File: src-java/flowmonitoring-topology/flowmonitoring-storm-topology/src/main/java/org/openkilda/wfm/topology/flowmonitoring/bolts/IslCacheBolt.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (rozdy) [reply src-java/flowmonitoring-topology/flowmonitoring-storm-topology/src/main/java/org/openkilda/wfm/topology/flowmonitoring/bolts/IslCacheBolt.java:71]: Fixed.

File: src-java/flowmonitoring-topology/flowmonitoring-storm-topology/src/main/java/org/openkilda/wfm/topology/flowmonitoring/bolts/IslCacheBolt.java [hunk 13]
```diff
@@ -0,0 +1,128 @@
...
+                } else {
+                    unhandledInput(input);
+                }
+            } else {
+                unhandledInput(input);
+            }
+            return;
+        }
+
+        if (ComponentId.FLOW_CACHE_BOLT.name().equals(input.getSourceComponent())) {
```
[the remark this answers is not in the document]
Comment (rozdy) [reply src-java/flowmonitoring-topology/flowmonitoring-storm-topology/src/main/java/org/openkilda/wfm/topology/flowmonitoring/bolts/IslCacheBolt.java:111]: Fixed.

**Review by dpoltavets — APPROVED** [review]

**Review by niksv — CHANGES_REQUESTED** [review]

File: confd/templates/flowmonitoring-topology/flowmonitoring-topology.tmpl [hunk 14]
```diff
@@ -0,0 +1,16 @@
...
+spouts:
+  - id: "coordinator.spout"
+    parallelism: 1
+  - id: "zookeeper.spout"
+    parallelism: 1
+
+# bolt definitions
+bolts:
+  - id: "zookeeper.bolt"
+    parallelism: 1
```
[the remark this answers is not in the document]
Comment (rozdy) [reply confd/templates/flowmonitoring-topology/flowmonitoring-topology.tmpl:16]: Added.

File: src-java/flowmonitoring-topology/flowmonitoring-storm-topology/src/main/java/org/openkilda/wfm/topology/flowmonitoring/FlowMonitoringTopology.java [hunk 15]
```diff
@@ -0,0 +1,178 @@
...
+                .allGrouping(ComponentId.ISL_LATENCY_SPOUT.name())
+                .allGrouping(ComponentId.ISL_SPOUT.name())
+                .allGrouping(ComponentId.FLOW_CACHE_BOLT.name())
+                .allGrouping(ZooKeeperSpout.SPOUT_ID);
+    }
+
+    private void actionBolt(TopologyBuilder topologyBuilder) {
+        declareBolt(topologyBuilder, new ActionBolt(), ComponentId.ACTION_BOLT.name())
+                .allGrouping(ComponentId.FLOW_CACHE_BOLT.name(), ACTION_STREAM_ID.name())
+                .allGrouping(ComponentId.ISL_CACHE_BOLT.name(), ACTION_STREAM_ID.name());
```
[the remark this answers is not in the document]
Comment (rozdy) [reply src-java/flowmonitoring-topology/flowmonitoring-storm-topology/src/main/java/org/openkilda/wfm/topology/flowmonitoring/FlowMonitoringTopology.java]: Fixed.

**Review by niksv — APPROVED** [review]

File: src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/fsm/create/FlowCreateFsm.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (timofei-durakov) [reply src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/fsm/create/FlowCreateFsm.java:474]: it's ok as for me to have to notify states, as long as these states are use the same action class

[the remark this answers is not in the document]
Comment (rozdy) [reply src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/fsm/create/FlowCreateFsm.java]: Reworked.

**Review by timofei-durakov — APPROVED** [review]

File: src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/service/FlowGenericCarrier.java [hunk 16]
```diff
@@ -58,4 +59,10 @@ default void sendActivateFlowMonitoring(RequestedFlow flow) {}
     default void sendDeactivateFlowMonitoring(String flow, SwitchId srcSwitchId, SwitchId dstSwitchId) {}
 
     void sendInactive();
+
+    /**
+     * Sends UpdateFlowInfo to flow-monitoring topology.
+     * @param flowInfo message to send
+     */
+    default void sendNotifyFlowMonitor(UpdateFlowInfo flowInfo) {}
```
Comment (sergii-iakovenko) [inline src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/service/FlowGenericCarrier.java:67]: What is the reason to have empty default implementation? Why not require from an every flow operation to notify the monitoring?

File: src-java/flowmonitoring-topology/flowmonitoring-storm-topology/src/main/java/org/openkilda/wfm/topology/flowmonitoring/bolts/ActionBolt.java [hunk 17]
```diff
@@ -0,0 +1,60 @@
...
+            Long maxLatencyTier2 = pullValue(input, MAX_LATENCY_TIER_2_FIELD, Long.class);
+
+            actionService.checkFlowSla(flowId, direction, latency, maxLatency, maxLatencyTier2);
+        } else {
+            unhandledInput(input);
+        }
+    }
+
+    @Override
+    public void declareOutputFields(OutputFieldsDeclarer declarer) {
```
Comment (sergii-iakovenko) [inline src-java/flowmonitoring-topology/flowmonitoring-storm-topology/src/main/java/org/openkilda/wfm/topology/flowmonitoring/bolts/ActionBolt.java:58]: What is the reason to override it with no implementation? Why not just keep as is?

File: src-java/flowmonitoring-topology/flowmonitoring-storm-topology/src/main/java/org/openkilda/wfm/topology/flowmonitoring/bolts/FlowCacheBolt.java [hunk 18]
```diff
@@ -0,0 +1,112 @@
...
+
+    public FlowCacheBolt(PersistenceManager persistenceManager, long flowRttStatsExpirationTime,
+                         String lifeCycleEventSourceComponent) {
+        super(lifeCycleEventSourceComponent);
+        this.persistenceManager = persistenceManager;
+        this.flowRttStatsExpirationTime = flowRttStatsExpirationTime;
+    }
+
+    @PersistenceContextRequired(requiresNew = true)
+    protected void init() {
```
Comment (sergii-iakovenko) [inline src-java/flowmonitoring-topology/flowmonitoring-storm-topology/src/main/java/org/openkilda/wfm/topology/flowmonitoring/bolts/FlowCacheBolt.java:65]: Did you forget "@Override" annotation here?

File: src-java/flowmonitoring-topology/flowmonitoring-storm-topology/src/main/java/org/openkilda/wfm/topology/flowmonitoring/bolts/IslCacheBolt.java [hunk 19]
```diff
@@ -0,0 +1,133 @@
...
+
+    public IslCacheBolt(PersistenceManager persistenceManager, long islRttLatencyExpiration,
+                        String lifeCycleEventSourceComponent) {
+        super(lifeCycleEventSourceComponent);
+        this.persistenceManager = persistenceManager;
+        this.islRttLatencyExpiration = islRttLatencyExpiration;
+    }
+
+    @PersistenceContextRequired(requiresNew = true)
+    protected void init() {
```
Comment (sergii-iakovenko) [inline src-java/flowmonitoring-topology/flowmonitoring-storm-topology/src/main/java/org/openkilda/wfm/topology/flowmonitoring/bolts/IslCacheBolt.java:66]: Did you forget "@Override" annotation here?

File: src-java/flowmonitoring-topology/flowmonitoring-storm-topology/src/main/java/org/openkilda/wfm/topology/flowmonitoring/mapper/FlowMapper.java [hunk 20]
```diff
@@ -0,0 +1,87 @@
...
+import org.mapstruct.ReportingPolicy;
+import org.mapstruct.factory.Mappers;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.stream.Collectors;
+
+@Mapper(unmappedTargetPolicy = ReportingPolicy.IGNORE)
+public interface FlowMapper {
```
Comment (sergii-iakovenko) [inline src-java/flowmonitoring-topology/flowmonitoring-storm-topology/src/main/java/org/openkilda/wfm/topology/flowmonitoring/mapper/FlowMapper.java:38]: It seems that the mapper operates with flowpath and segments only. Why not call it FlowPathMapper?

File: src-java/flowmonitoring-topology/flowmonitoring-storm-topology/src/main/java/org/openkilda/wfm/topology/flowmonitoring/service/ActionService.java [hunk 21]
```diff
@@ -0,0 +1,38 @@
...
+@Slf4j
+public class ActionService {
+
+    /**
+     * Check flow SLA is violated.
+     */
+    public void checkFlowSla(String flowId, FlowDirection direction, long latency,
+                             Long maxLatency, Long maxLatencyTier2) {
+        if (maxLatency != null && latency > maxLatency) {
+            long latencyMs = latency / 1000;
```
Comment (sergii-iakovenko) [inline src-java/flowmonitoring-topology/flowmonitoring-storm-topology/src/main/java/org/openkilda/wfm/topology/flowmonitoring/service/ActionService.java:31]: Why not use TimeUnit?
